run_squad.py

```diff
@@ -1141,15 +1152,32 @@ def main(_):
         FLAGS.tpu_name, zone=FLAGS.tpu_zone, project=FLAGS.gcp_project)
 
   is_per_host = tf.contrib.tpu.InputPipelineConfig.PER_HOST_V2
-  run_config = tf.contrib.tpu.RunConfig(
-      cluster=tpu_cluster_resolver,
-      master=FLAGS.master,
-      model_dir=FLAGS.output_dir,
-      save_checkpoints_steps=FLAGS.save_checkpoints_steps,
-      tpu_config=tf.contrib.tpu.TPUConfig(
-          iterations_per_loop=FLAGS.iterations_per_loop,
-          num_shards=FLAGS.num_tpu_cores,
-          per_host_input_for_training=is_per_host))
+
+  if FLAGS.use_multi_gpu:
+    # Horovod: pin GPU to be used to process local rank (one GPU per process)
+    config = tf.ConfigProto()
+    config.gpu_options.visible_device_list = str(hvd.local_rank())
+    run_config = tf.contrib.tpu.RunConfig(
+        cluster=tpu_cluster_resolver,
+        master=FLAGS.master,
+        model_dir=FLAGS.output_dir,
+        save_checkpoints_steps=FLAGS.save_checkpoints_steps,
+        tpu_config=tf.contrib.tpu.TPUConfig(
+            iterations_per_loop=FLAGS.iterations_per_loop,
+            num_shards=FLAGS.num_tpu_cores,
```

Comment:
Is num_shards still used here? TPUConfig seems to have deprecated this one.
Also, do we need d=d.shard(hvd.size(), hvd.rank())?

Reply:
This PR was tested with TF 1.12. You are right and TF 2.X doc says:

> num_shards: (Deprecated, ignored by TPUEstimator). The number of model replicas in the system. For non-model-parallelism case, this number equals the total number of TPU cores. For model-parallelism, the total number of TPU cores equals num_cores_per_replica * num_shards.

Regarding the data sharding, all the workers randomly sample the data and shuffle, so there is no need to shard the data.

Reply:
Thank you @abditag2 for the explanation. 
Have you by any chance tested BERT with TF 2.x and Horovod? What changes will you be making if you were to enable TF 2.x, esp the above shard related code, any inputs are welcome, thank you! :)

Reply:
Horovod itself if compatible with TF 2.x (we test it with nightly builds to ensure that). 

From the source code and comments (https://github.com/tensorflow/estimator/blob/e2e2b5e19f7dcdea05c27280f9cfa72f4255529e/tensorflow_estimator/python/estimator/tpu/tpu_config.py#L78), if you remove the num_shards and have `num_cores_per_replica` be `None`, you would end up with disabling TF's model parallelism which is what you want when using Horovod. When running a TPUEstimator on a machine without TPUs, the default TF behavior is to run it on the GPU and you want single model per GPU and then use horovod for gradient aggregation.

You have to do this only in the multiGPU mode because you still want the regular TPU config when using TPUs.

Reply:
Ok, thank you @abditag2!